contrib/modelci/README.md

```diff
@@ -0,0 +1,143 @@
+        key = factor.__name__
+        assert key not in Factor.dic
+        Factor.dic[key] = factor
+    
+       def __del__(self):
+        if self.factors:
+            # write to file self.out_file
+```
+
+More factors can be integrated into the test framework, for example, a factor tracker which test the training duration can be added in the following way
+
+```python
+class TrainDurationFactor(Factor):
+    def __init__(self, threshold):
+        super(TrainDurationFactor, self).__init__('train.dura.txt')
+        self.threshold = threshold
+    
+    def test(self):
+        cur_data = _load_nparray_from_file(self.out_file)
+        his_data = _load_nparray_from_file(self.his_file)
+        diff = np.abs(cur_data - his_data) / his_data
+        if (diff > self.threshold).any():
+            raise TestError
+# register this factor tracker into the framework      
+Factor.register(TrainDurationFactor)
+```
+
+A testable model should have a file called `meta.py` with some configurations about those factors to use like
+
+```python
+# this is a demo for meta.py
+train_duration_factor = TrainDurationFactor(0.1)
+valid_duration_factor = ValidDurationFactor()
+train_memory_factor = TrainMemoryFactor()
+
+tracking_factors = [ train_duration, valid_duration_factor, train_meta_factor ]
+```
+
+Inside the model, one should call the factor trackers and add records.
+
+```python
+# configuration of some model
+import paddle
+import meta
+# ...
+
+
+for batch in batches:
+    # ...
+    duration = _get_duration_of_this_batch()
+    train_duration_factor.add_record(duration)
+    # ...
+```
+
+and the test framework will test each factor like
+
+```python
+for tracker in some_model.meta.tracking_factors:
+    some_model._run() # run and tracking_factors will collect running status
+    try:
+        tracker.test()
+    except TestError:
+        _collect_error_info
+        _ring_alarm
+```
+
+## Build a testable model
+
+The models should be placed in `./models` directory, each has a sub-directory, and a `train.xsh` script to define how to run this model. After triggering the `train.xsh`, all the data of `tracking_factors` should be created.
+
+For example, a normal model might have following logic
+
+```python
+# train.xsh
+run_train_cpu
+run_train_gpu
+```
+
+To make the testing logic stable, the testable model should assure that
+
+- fix the random seed to make result reproducible
+- just run 10 or 20 batches, and each batch takes no more than 30 seconds to limit the overall execution time
+- run different modes sequentially, not run them parallelly
+
+## Persistence of log
+
+The log of each execution should be stored somewhere, 
+the simplest way is to use Git to maintain a versionable history.
+
+After each execution, add all the logs and statistic result and commit with a comment with a 
+template like
+
+```
+{success_or_not} {short summary of the cause}
+
+execution duration: {overall_duration}
+paddle code version: {commitid}
```

Comment:
How about using a proto text format? In the future, we might want a dashboard to read the log and visualize it.

Reply:
Agree, proto can store more complex data structure, but a plain text file with some specific file name seems enough to store the values of most KPI. 

For example, to store the data for `train cost`, a file called `./train.cost.txt` will be created with data as follows
```
1.2
0.99
0.4
```
each line a float number, and not hard to load and visualize.

I wonder whether there is some KPI that has a complex data structure so that a plain text is not enough to store the data.

Another question, I am confused about how to persist the history data, a Github repo is good for version and a plain text is human-friendly for analysis, but it can not scale up.

A database is good for scale-up, but works like a black box and make the need for the ModelCI to support some analysis accessing data from the black box.
@panyx0718

Reply:
I think we can first save it to local disk. In the future, we might apply for some storage resource in the cloud.